File: overlay/clojure/hydra-kernel/src/main/clojure/hydra/overlay/clojure/lib/system.clj
(ns hydra.overlay.clojure.lib.system
  (:require [hydra.system :refer :all]
            [hydra.time :refer :all])
  (:import [java.io ByteArrayOutputStream IOException InputStream File]))

;; Clojure implementations of hydra.lib.system primitives (#498).
;;
;; The Hydra type effect<t> is transparent in the Lisp dialects (effect<t> = t), so these helpers
;; perform their work eagerly. Fallible primitives return a Hydra Either: success is (list :right v),
;; a launch failure is (list :left error). Mirrors the classify pattern of the Haskell/Java/Scala hosts.
;;
;; Runtime representations (Clojure target):
;;   Either              : (list :left v) | (list :right v)
;;   Optional            : (list :given v) | (list :none)
;;   SystemError         : (list :<variant> payload), variant one of
;;                         :command_not_found | :permission_denied | :invalid_working_directory
;;                         | :interrupted | :other; payload a FilePath (string) or message string,
;;                         and () for the nullary :interrupted.
;;   Command             : a hydra_system_command defrecord {:program :arguments :working_directory :environment}
;;   ProcessResult       : a hydra_system_process_result defrecord {:exit_code :stdout :stderr}
;;   FilePath/StatusCode/EnvironmentVariable : transparent (bare string / int / string)
;;   binary              : a vector of ints 0..255 (matches hydra.overlay.clojure.lib.files / literals)
;;   unit                : nil
;;
;; All functions are curried, matching the Clojure prim runtime style.

;; ---- Helpers (not primitives) ----

(defn- message [^Exception e]
  (let [m (.getMessage e)]
    (if (nil? m) (.getSimpleName (class e)) m)))

;; Java byte-array -> binary (vector of unsigned ints 0..255).
(defn- bytes->binary [^bytes ba]
  (vec (map #(bit-and % 0xff) ba)))

;; Read an InputStream fully into a Java byte-array.
(defn- read-all [^InputStream in]
  (let [buffer (ByteArrayOutputStream.)
        chunk (byte-array 8192)]
    (loop []
      (let [n (.read in chunk)]
        (when (not= n -1)
          (.write buffer chunk 0 n)
          (recur))))
    (.toByteArray buffer)))

;; Classify a process-launch IOException into a SystemError.
(defn- classify [program ^IOException e]
  (let [lower (.toLowerCase (message e))]
    (cond
      (or (.contains lower "error=2") (.contains lower "no such file"))     (list :command_not_found program)
      (or (.contains lower "error=13") (.contains lower "permission denied")) (list :permission_denied program)
      (or (.contains lower "error=20") (.contains lower "not a directory"))  (list :invalid_working_directory program)
      :else                                                                  (list :other (message e)))))

;; ---- Primitives ----

;; execute :: Command -> effect<Either<SystemError, ProcessResult>>
(def hydra_lib_system_execute
  "Run a program to completion, capturing stdout/stderr (binary) and the exit code."
  (fn [command]
    (let [program (:program command)
          args (:arguments command)
          wd (:working_directory command)
          env (:environment command)
          argv (java.util.ArrayList. ^java.util.Collection (cons program args))
          pb (ProcessBuilder. argv)]
      (when (= (first wd) :given)
        (.directory pb (File. ^String (second wd))))
      (when (= (first env) :given)
        (let [pe (.environment pb)]
          (.clear pe)
          (doseq [[k v] (second env)] (.put pe k v))))
      (try
        (let [process (.start pb)
              out (read-all (.getInputStream process))
              err (read-all (.getErrorStream process))
              code (.waitFor process)]
          (list :right (->hydra_system_process_result code (bytes->binary out) (bytes->binary err))))
        (catch IOException e (list :left (classify program e)))
        (catch InterruptedException _e
          (.interrupt (Thread/currentThread))
          (list :left (list :interrupted)))))))

;; exit :: StatusCode -> effect<unit>
(def hydra_lib_system_exit
  "Terminate the current process immediately with the given status. Does not return."
  (fn [code]
    (System/exit code)
    nil))

;; getEnvironment :: effect<Map<EnvironmentVariable, string>>
;; A nullary effect: generated consumer code references this as a bare value (the effect result), so it
;; is a plain def, not a thunk. (The reducer registry wraps it in a thunk; see register-system.)
(def hydra_lib_system_get_environment
  "Return the entire environment of the current process as a map from variable name to value."
  (into {} (System/getenv)))

;; getEnvironmentVariable :: EnvironmentVariable -> effect<Optional<string>>
(def hydra_lib_system_get_environment_variable
  "Return the value of the named environment variable, or none if it is not set."
  (fn [name]
    (let [v (System/getenv ^String name)]
      (if (nil? v) (list :none) (list :given v)))))

;; getTime :: effect<Timespec>  (nullary effect: a bare value, see getEnvironment note)
(def hydra_lib_system_get_time
  "Return the current wall-clock time as a Timespec (seconds and nanoseconds since the Unix epoch)."
  (let [now (java.time.Instant/now)]
    (->hydra_time_timespec (.getEpochSecond now) (long (.getNano now)))))

;; getWorkingDirectory :: effect<Either<SystemError, FilePath>>  (nullary effect: a bare value)
(def hydra_lib_system_get_working_directory
  "Return the current working directory as a FilePath (string)."
  (let [cwd (System/getProperty "user.dir")]
    (if (nil? cwd)
      (list :left (list :other "user.dir is not set"))
      (list :right cwd))))

File: overlay/typescript/hydra-kernel/src/main/typescript/hydra/overlay/typescript/lib/system.ts
// Hand-written runtime: hydra.lib.system primitives.
//
// Real system-interface primitives via synchronous Node.js APIs. The effect
// type is transparent in TypeScript (effect<t> = t), so these perform their
// effect eagerly. Fallible primitives return Either<SystemError, T>: a failure
// to launch (or perform the call) is Left(classified SystemError); success is
// Right(value). Infallible primitives (getEnvironment, getEnvironmentVariable,
// getTime) return their value directly. Mirrors the Python reference in
// overlay/python/hydra-kernel/.../hydra/python/lib/system.py and the Haskell
// reference Hydra.Haskell.Lib.System. For #498.
//
// `binary` values (ProcessResult.stdout/stderr) are base64-encoded strings (the
// TS runtime convention from literals.ts / files.ts), so they pair with
// hydra.lib.text.decodeUtf8. `SystemError`, `Command`, `ProcessResult`, etc. are
// generated types; we use `any` here to stay decoupled from the generated dist.

import { spawnSync } from "node:child_process";
import process from "node:process";
import type { Either, Optional } from "../../../runtime.js";
import { Left, Right, Given, None } from "../../../runtime.js";

// ---- Helpers ---------------------------------------------------------------

function bufferToBase64(buf: Buffer | Uint8Array | null | undefined): string {
  return Buffer.from(buf ?? Buffer.alloc(0)).toString("base64");
}

// Classify a spawn launch failure (errno) into a SystemError tagged variant.
function classifyLaunch(program: any, e: unknown): any {
  const code = (e as NodeJS.ErrnoException).code;
  if (code === "ENOENT")                       return { tag: "commandNotFound", value: program };
  if (code === "EACCES" || code === "EPERM")   return { tag: "permissionDenied", value: program };
  if (code === "ENOTDIR")                      return { tag: "invalidWorkingDirectory", value: program };
  if (code === "EINTR")                        return { tag: "interrupted" };
  return { tag: "other", value: e instanceof Error ? e.message : String(e) };
}

// ---- Primitives ------------------------------------------------------------

// execute : Command -> effect<either<SystemError, ProcessResult>>
// Runs a program to completion and captures its result. A child that runs and
// exits non-zero is Right(result) with that exitCode; only a launch failure is
// Left(error). No shell is invoked; the program is executed directly.
export const execute = (command: any): Either<any, any> => {
  const program: string = command.program.value;
  const args: string[] = Array.from(command.arguments_ ?? []);
  const cwd = command.workingDirectory?.tag === "given"
    ? command.workingDirectory.value.value
    : undefined;
  let env: Record<string, string> | undefined = undefined;
  if (command.environment?.tag === "given") {
    env = {};
    for (const [k, v] of command.environment.value) env[k.value] = v;
  }
  const r = spawnSync(program, args, {
    cwd,
    env,
    shell: false,
    windowsHide: true,
  });
  if (r.error) {
    return Left(classifyLaunch(command.program, r.error));
  }
  return Right({
    exitCode: { value: r.status ?? 0 },
    stdout: bufferToBase64(r.stdout),
    stderr: bufferToBase64(r.stderr),
  });
};

// exit : StatusCode -> effect<unit>
// Terminate the current process with the given status code. Does not return.
export const exit = (code: any): never => {
  process.exit(code.value);
};

// getEnvironment : effect<map<EnvironmentVariable, string>>
// The full environment, as a map from variable name to value.
export const getEnvironment = (): ReadonlyMap<any, string> => {
  const m = new Map<any, string>();
  for (const [k, v] of Object.entries(process.env)) {
    if (v !== undefined) m.set({ value: k }, v);
  }
  return m;
};

// getEnvironmentVariable : EnvironmentVariable -> effect<optional<string>>
// Look up a single environment variable by name; none if it is not set.
export const getEnvironmentVariable = (name: any): Optional<string> => {
  const value = process.env[name.value];
  return value === undefined ? None : Given(value);
};

// getTime : effect<Timespec>
// The current wall-clock time as a Timespec (seconds and nanoseconds since the
// Unix epoch). Both seconds and nanoseconds are bigint, matching the generated
// Timespec type (int64 and uint32 both encode as bigint in TypeScript).
export const getTime = (): any => {
  const nanosTotal = BigInt(Math.round(Date.now())) * 1_000_000n;
  // Date.now() gives wall-clock ms since epoch; convert to seconds + nanoseconds.
  const seconds = nanosTotal / 1_000_000_000n;
  const nanoseconds = nanosTotal % 1_000_000_000n;
  return { seconds, nanoseconds };
};

// getWorkingDirectory : effect<either<SystemError, FilePath>>
// The current working directory as a FilePath.
export const getWorkingDirectory = (): Either<any, any> => {
  try {
    return Right({ value: process.cwd() });
  } catch (e) {
    return Left({ tag: "other", value: e instanceof Error ? e.message : String(e) });
  }
};
